File: backend/app/domain/entities/efficiency.py
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, Field


class PeriodData(BaseModel):
    """Period data for efficiency calculations"""
    
    period: str = Field(..., description="Period name (e.g., 'business_hours', 'after_hours', 'weekend')")
    time_range: str = Field(..., description="Time range (e.g., '08:00-18:00')")
    days: List[str] = Field(..., description="Days of the week")
    current_electric_kwh: float = Field(..., gt=0, description="Current electric consumption in kWh")
    current_gas_therms: float = Field(..., gt=0, description="Current gas consumption in therms")
    baseline_electric_kwh: float = Field(..., gt=0, description="Baseline electric consumption in kWh")
    baseline_gas_therms: float = Field(..., gt=0, description="Baseline gas consumption in therms")
    electric_rate: float = Field(..., gt=0, description="Electric rate per kWh")
    gas_rate: float = Field(..., gt=0, description="Gas rate per therm")


class PeriodEfficiencyMetrics(BaseModel):
    """Efficiency metrics for a specific period"""
    
    period: str
    time_range: str
    days: List[str]
    
    # Energy savings
    electric_savings_kwh: float = Field(..., description="Electric energy savings in kWh")
    gas_savings_therms: float = Field(..., description="Gas energy savings in therms")
    
    # Cost savings
    electric_cost_savings: float = Field(..., description="Electric cost savings")
    gas_cost_savings: float = Field(..., description="Gas cost savings")
    total_cost_savings: float = Field(..., description="Total cost savings")
    
    # Efficiency improvements
    electric_efficiency_improvement: float = Field(..., description="Electric efficiency improvement percentage")
    gas_efficiency_improvement: float = Field(..., description="Gas efficiency improvement percentage")
    overall_efficiency_improvement: float = Field(..., description="Overall efficiency improvement percentage")


class EfficiencySummary(BaseModel):
    """Summary of efficiency calculations"""
    
    total_electric_savings_kwh: float = Field(..., description="Total electric savings in kWh")
    total_gas_savings_therms: float = Field(..., description="Total gas savings in therms")
    total_electric_cost_savings: float = Field(..., description="Total electric cost savings")
    total_gas_cost_savings: float = Field(..., description="Total gas cost savings")
    total_cost_savings: float = Field(..., description="Total cost savings")
    average_electric_efficiency_improvement: float = Field(..., description="Average electric efficiency improvement")
    average_gas_efficiency_improvement: float = Field(..., description="Average gas efficiency improvement")
    overall_efficiency_improvement: float = Field(..., description="Overall efficiency improvement")
    performance_grade: str = Field(..., description="Performance grade (A, B, C, D, F)")


class EfficiencyCalculation(BaseModel):
    """Efficiency calculation domain entity"""
    
    id: Optional[str] = None
    building_id: str = Field(..., description="Building identifier")
    measure_name: str = Field(..., description="Energy efficiency measure name")
    calculation_timestamp: datetime = Field(default_factory=datetime.utcnow, description="When the calculation was performed")
    periods: List[PeriodEfficiencyMetrics] = Field(..., description="Efficiency metrics for each period")
    summary: EfficiencySummary = Field(..., description="Aggregated summary of all periods")
    created_at: Optional[datetime] = None
    
    class Config:
        """Pydantic configuration"""
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class BuildingEfficiencySummary(BaseModel):
    """Building efficiency summary for GET endpoints"""
    
    building_id: str
    total_calculations: int
    latest_calculation: Optional[EfficiencyCalculation] = None
    best_performance_grade: Optional[str] = None
    average_efficiency_improvement: Optional[float] = None
    total_cost_savings: Optional[float] = None
    created_at: Optional[datetime] = None
    
    class Config:
        """Pydantic configuration"""
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

File: frontend/types/v0/efficiency.ts
export interface PeriodData {
  period: string
  time_range: string
  days: string[]
  current_electric_kwh: number
  current_gas_therms: number
  baseline_electric_kwh: number
  baseline_gas_therms: number
  electric_rate: number
  gas_rate: number
}

export interface CalculateEfficiencyRequest {
  building_id: string
  measure_name: string
  periods: PeriodData[]
}

export interface PeriodEfficiencyMetrics {
  period: string
  time_range: string
  days: string[]
  electric_savings_kwh: number
  gas_savings_therms: number
  electric_cost_savings: number
  gas_cost_savings: number
  total_cost_savings: number
  electric_efficiency_improvement: number
  gas_efficiency_improvement: number
  overall_efficiency_improvement: number
}

export interface EfficiencySummary {
  total_electric_savings_kwh: number
  total_gas_savings_therms: number
  total_electric_cost_savings: number
  total_gas_cost_savings: number
  total_cost_savings: number
  average_electric_efficiency_improvement: number
  average_gas_efficiency_improvement: number
  overall_efficiency_improvement: number
  performance_grade: string
}

export interface EfficiencyCalculation {
  id: string
  building_id: string
  measure_name: string
  calculation_timestamp: string
  periods: PeriodEfficiencyMetrics[]
  summary: EfficiencySummary
  created_at: string
}

export interface BuildingEfficiencySummary {
  building_id: string
  total_calculations: number
  latest_calculation?: EfficiencyCalculation
  best_performance_grade?: string
  average_efficiency_improvement?: number
  total_cost_savings?: number
  created_at: string
}

export interface BuildingCalculationsResponse {
  building_id: string
  calculations: EfficiencyCalculation[]
  total_count: number
}

export interface AllBuildingsSummaryResponse {
  buildings: BuildingEfficiencySummary[]
  total_buildings: number
  page: number
  limit: number
  total_pages: number
  has_next: boolean
  has_prev: boolean
}

// Form data types
export interface EfficiencyFormData {
  building_id: string
  measure_name: string
  periods: PeriodFormData[]
}

export interface PeriodFormData {
  period: string
  time_range: string
  days: string[]
  current_electric_kwh: number
  current_gas_therms: number
  baseline_electric_kwh: number
  baseline_gas_therms: number
  electric_rate: number
  gas_rate: number
}

// Performance grade colors
export const PERFORMANCE_GRADE_COLORS = {
  A: "text-green-600 bg-green-50 border-green-200",
  B: "text-blue-600 bg-blue-50 border-blue-200",
  C: "text-yellow-600 bg-yellow-50 border-yellow-200",
  D: "text-orange-600 bg-orange-50 border-orange-200",
  F: "text-red-600 bg-red-50 border-red-200",
} as const

export type PerformanceGrade = keyof typeof PERFORMANCE_GRADE_COLORS
